### PR TITLE
flake.nix: provide checks (tests) and docs and related fixes

### LINKED ISSUE #2722: bug: attribute 'shellDryRun' missing in 'modules/programs/bash.nix'
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

With the current update, the bash module in home-manager started breaking for me. I did not change anything in the program.bash set, but started getting the following error:

```
error: attribute 'shellDryRun' missing, at /home/fabian/.nix-defexpr/channels/home-manager/modules/programs/bash.nix:12:9
```

The issue persists after updating my user nixpkgs channel to the version stated below. 
This seems to be related to #2693 and #2668.

### Maintainer CC

@rycee

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 5.10.90, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.335130.386234e2a61"`
 - channels(fabian): `"home-manager, nixos-21.11.335946.c28fb0a4671, nixpkgs-unstable-22.05pre352484.1882c6b7368"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by ShamrockLee**
### Description

This patch provides the Home Manager Module test cases from flake through `outputs.checks.${system}` and `outputs.legacyPackages.${system}`. This allows developers to run the tests against the master branch without having to update their environment.

If applied, developers can run the tests by
```
nix develop --override-input nixpkgs github:NixOS/nixpkgs/nixos-unstable .#tests.run.all
```
~or simply~
```
nix flake check --override-input nixpkgs github:NixOS/nixpkgs/nixos-unstable .
```

Update: As reported by https://github.com/nix-community/home-manager/pull/3024#issuecomment-1156281495 , the multi-platform failure would revive on 22.05 even after the IFD fix. We should probably add the test suite into the `checks` until https://github.com/NixOS/nix/issues/4265 got fixed.

(`nix develop` is the Flake analogy of `nix-shell --pure -A`. The `--override-input` above makes Nix build against the Nixpkgs nixos-unstable branch instead of the master branch to avoid rebuilding the world.)

Fixes #2722 when running the tests for flake users.

This patch also contains two fixes found when testing the above features:
*   Fix impurities when evaluating `tests` introduced in `modules/modules.nix` by the appended element `pkgsModule` that injects `_module.args.pkgsPath` as `<nixpkgs>` .
*   Get `pkgs` through `inputs.nixpkgs.legacyPackages.${system}` whenever available instead of  `import nixpkgs { inherit "${system}"; }`
    *    ~This avoids IFD where possible, working around https://github.com/NixOS/nix/issues/4265 on platforms covered by `eachDefaultSystem`.~ Update: not on NixOS 22.05.
    *     This prevents re-evaluation of Nixpkgs. See [1000 instances of nixpkgs](https://discourse.nixos.org/t/1000-instances-of-nixpkgs/17347) for detail.

(I don't know why the last sub-bullet is rendered as a code block. It's probably a GitHub bug.

Regarding #3056:
This PR does not mean to revert the decisions made in the PR made by rycee. Rycee's PR cleared up the hack to use flake.lock content on non-flake system, with the side effect of removing the flake support to access the documents. This PR adds back the flake-based document support plus the tests support, while paying respect to the non-flake infrastructure.

### Checklist

- [X] Change is backwards compatible. (I think so, but the impurity fix requires experienced people to double check.)

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.
    -   Tested with both ways mentioned above.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by ncfavier** [comment]
This still runs into https://github.com/NixOS/nix/issues/4265: if I run just `nix flake check` I get
```
error: a 'aarch64-linux' with features {} is required to build '/nix/store/fijjj1wrarrxbs019dibrg6c8m11zzz3-assertions.expected.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, ca-derivations, kvm, nixos-test}
```

Also I think there might be a clash between how flake checks work (you build a derivation, the build succeeds iff the test succeeds) vs how NMT works (tests are meant to be run in nix-shell).

**Comment by ShamrockLee** [comment]
> This still runs into [NixOS/nix#4265](https://github.com/NixOS/nix/issues/4265): if I run just `nix flake check` I get
> 
> ```
> error: a 'aarch64-linux' with features {} is required to build '/nix/store/fijjj1wrarrxbs019dibrg6c8m11zzz3-assertions.expected.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, ca-derivations, kvm, nixos-test}
> ```
> 
> Also I think there might be a clash between how flake checks work (you build a derivation, the build succeeds iff the test succeeds) vs how NMT works (tests are meant to be run in nix-shell).

Sounds like something happens between Nixpkgs 21.11 and 22.05. I'm currently bothered by some failure that looks like https://github.com/NixOS/nixpkgs/issues/171281 that occurs after upgrading to 22.05, so I'm still using 21.11.

I'll remove the `checks` output then.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ Also make sure to read the guidelines found at
 
 - [ ] Code formatted with `./format`.
 
-- [ ] Code tested through `nix-shell --pure tests -A run.all`.
+- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --override-input nixpkgs github:NixOS/nixpkgs/nixos-unstable .#run.all`.
```
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: I didn't notice we have a lock file now. (It used to be `.gitignore`d.)

[the remark this answers is not in the document]
Comment (ncfavier) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: See https://github.com/nix-community/home-manager/pull/1934 for discussion

File: flake.nix [hunk 2]
```diff
@@ -46,11 +46,14 @@
...
       let
-        pkgs = import nixpkgs { inherit system; };
+        pkgs = nixpkgs.legacyPackages.${system} or (import nixpkgs {
+          inherit system;
+        });
         docs = import ./docs {
           inherit pkgs;
           nmdSrc = nmd;
         };
+        tests = import ./tests { inherit pkgs; };
```
[the remark this answers is not in the document]
Comment (ncfavier) [reply flake.nix:54]: Yes

**Comment by ShamrockLee** [comment]
Rebase and update after #2722.

This PR does not mean to revert the decisions made in rycee's PR. Rycee's PR cleared up the hack to use flake.lock content on non-flake system, with the side effect of removing the flake support to access the documents. This PR adds back the flake-based document support plus the tests support, while paying respect to the non-flake infrastructure.

File: tests/default.nix [hunk 3]
```diff
@@ -1,20 +1,27 @@
...
     inherit lib pkgs;
     check = false;
   } ++ [
     {
+      # Fix impurities.
+      # The last element of the above `modules` (i.e. `pkgsModule`)
+      # would set `pkgsPath` with `mkDefault` to `<nixpkgs>`
+      # if `config.home.stateVersion` is before 20.09
+      _module.args.pkgsPath = pkgs.path;
+
```
Comment (berbiche) [inline tests/default.nix:24]: Not sure if this should be in an external PR

**Review by berbiche — APPROVED** [review]
LGTM besides one thing I'm not sure of.

**Comment by ncfavier** [comment]
I think this goes in the right direction, but now we have *two* sources for both nmd and nmt, and twice as much work needed to update them. If we're going to have nmd and nmt as flake inputs, I really see no good reason to use `fetchTarball` rather than `flake-compat`.

@rycee are you still opposed to using flakes here, even after the comments in https://github.com/nix-community/home-manager/pull/3056#issuecomment-1166642374 (reading past the aggressive tone)?

**Comment by ShamrockLee** [comment]
I just noticed that the hashes of the tarball fetcheres are hard-coded. To keep such decision, I may just drop the flake inputs `nmd` and `nmt`. There would be a room for discussion as mentioned in the above comment.

**Comment by ShamrockLee** [comment]
I just noticed that the hashes of the tarball fetcheres are hard-coded.

To keep such decision, I may just drop the flake inputs `nmd` and `nmt`. There might be room for discussion as mentioned in the above comment.

File: tests/default.nix (hunk 3 above)
Comment (ShamrockLee) [reply tests/default.nix:24]: I'll split it out to make it merge faster, considering the discussion to be made about nmd and nmt.

**Comment by ShamrockLee** [comment]
@berbiche The impurity fix is now splitted as #3078. Please take a look.

**Comment by ShamrockLee** [comment]
I don't know what happened, I'll re-open the PR again.

**Pull request closed without merging**